### PR TITLE
Added native canvas drawing

### PR DESCRIPTION
**Pull request by joydeep-b**
Example usage:
```js
c = lib220.newCanvas(600, 600);
c.drawArc(300, 300, 200, 0, 2 * Math.PI, [0.8, 0.4, 0]);

for (let i = 0; i < 100; ++i) {
  const a = i / 100 * Math.PI;
  const x = 200 * Math.cos(a);
  const y = 200 * Math.sin(a);
  c.drawLine(300 + x, 300 + y, 300 - x,
    300 - y, [1, 0.5, 0]);
}
```

### CODE REVIEW

**Comment by arjunguha** [comment]
Yikes, why are you doing this? Should we consider using a 3rd party library or just using the canvas directly?

**Comment by joydeep-b** [comment]
This is using the canvas directly. It just abstracts away the notion of
paths and strokes, etc.

File: ts/lib220.ts [hunk 1]
```diff
@@ -1,5 +1,79 @@
...
+  constructor(w: number, h: number) {
+    if (arguments.length !== 2) {
+      throw new TypeError(`Failed to construct Node 'DrawingCanvas': 2 arguments required but ${arguments.length} given`);
+    }
+    this.width = w;
+    this.height = h;
+    if (typeof document === 'undefined') {
+      return; //  for node
+    }
+    const canvases = document.getElementById('canvases')!;
```
Comment (arjunguha) [inline ts/lib220.ts]: So, some weird effects will occur if a program chooses to show an image and use drawingCanvas at the same time. I think we should just be aware of it and totally ignore it for now. This exists only for in-class demos.

File: ts/lib220.ts [hunk 2]
```diff
@@ -1,5 +1,79 @@
...
+  for (let i = 0; i < 3; ++i) {
+    hex += hexColorChannel(rgb[i]);
+  }
+  return hex;
+}
+
+export class DrawingCanvas {
+  width: number = 1;
+  height: number = 1;
+  ctx: CanvasRenderingContext2D | undefined = undefined;
```
Comment (arjunguha) [inline ts/lib220.ts]: Just write ` ctx: CanvasRenderingContext2D;`

TypeScript will check the constructor and see that ctx is initialized (strict class initialization):

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html

Comment (joydeep-b) [reply ts/lib220.ts]: That won't work since `ctx` does not get set if this function is called under node.

File: ts/lib220.ts (hunk 1 above)
Comment (joydeep-b) [reply ts/lib220.ts]: The drawing actually happens in a saved context `ctx`, so creation of new canvases is not a problem - it still draws on the correct canvas (the one that was created in the constructor).

What effect were you thinking of? The one edge case that we have is that we delete older canvases. Interestingly, the browser happily accepts drawing commands to a context which no longer exists, without any errors.

Comment (arjunguha) [reply ts/lib220.ts]: I understand it will draw to the right canvas, but I was thinking about interleaving.

File: ts/lib220.ts (hunk 2 above)
Comment (arjunguha) [reply ts/lib220.ts]: OK, this is fine. We should just merge. But, if we want students to actually use this, the `newCanvas` function should return a mock object that silently fails.

Comment (joydeep-b) [reply ts/lib220.ts]: Oh, true. Done.

**Pull request merged**